src/librustc/traits/object_safety.rs

```diff
@@ -91,8 +91,10 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn astconv_object_safety_violations(self, trait_def_id: DefId)
                                             -> Vec<ObjectSafetyViolation>
     {
+        debug_assert!(self.generics_of(trait_def_id).has_self);
+        let self_ty = self.mk_self_type();
         let violations = traits::supertrait_def_ids(self, trait_def_id)
-            .filter(|&def_id| self.predicates_reference_self(def_id, true))
+            .filter(|&def_id| self.predicates_reference_self(def_id, self_ty, true))
```

Comment:
It should be possible to pre-intern the self type in `CommonTypes`.
In that case it won't have to be passed as an argument everywhere.